File: 0523-continuous-subarray-sum/0523-continuous-subarray-sum.py
class Solution:
    def checkSubarraySum(self, nums: List[int], k: int) -> bool:
        n = len(nums)
        psum = [0]
        ref = {0: 0}
        for i in range(n):
            psum.append(psum[-1] + nums[i])
            remainder = psum[i+1] % k
            if remainder in ref:
                if i > ref[remainder]:
                    return True
            else:
                ref[remainder] = i + 1

        return False

File: 0523-continuous-subarray-sum/0523-continuous-subarray-sum.cpp
class Solution {
public:
    bool checkSubarraySum(vector<int>& nums, int k) {
        unordered_map<int, int> ref;
        ref[0] = -1;
        int running = 0;

        for (int i = 0; i < nums.size(); i++) {
            running += nums[i];
            if (ref.contains(running % k)) {
                if (i - ref[running % k] > 1) return true;
            } else ref[running % k] = i;
        }

        return false;
    }
};